litestar/handlers/http_handlers/base.py

```diff
@@ -499,12 +497,15 @@ def _validate_handler_function(self) -> None:
                 "If the function should return a value, change the route handler status code to an appropriate value.",
             )
 
+        if (
+            before_request := self.resolve_before_request()
+        ) and not before_request.parsed_signature.return_type.is_optional:
+            return_type = before_request.parsed_signature.return_type
```

Comment:
Inspecting the return annotation of the `before_request` handler in order to determine what the _actual_ response of the handler is, is something that we do somewhere else also. However, I don't believe we take into account that the `before_request` handler may optionally return a response, in which case the true response type of the handler is the union of the non-optional component of the `before_request` handler, and the response of the handler itself.

Reply:
Here:
https://github.com/litestar-org/litestar/blob/f9952573ba3a0f7e5aa742682c437bf412f1b5ab/litestar/handlers/http_handlers/base.py#L420-L423

This is to determine which response handler should be used. I believe that if the `before_request` handler has a return annotation of `RedirectResponse | None`, then we probably don't end up with the right response handler in this case.